File: hu/codes/java/chapter_hashing/array_hash_map.java
/**
 * File: array_hash_map.java
 * Created Time: 2022-12-04
 * Author: krahets (krahets@163.com)
 */

package chapter_hashing;

import java.util.*;

/* Kulcs-érték pár */
class Pair {
    public int key;
    public String val;

    public Pair(int key, String val) {
        this.key = key;
        this.val = val;
    }
}

/* Tömb alapú hash tábla implementáció */
class ArrayHashMap {
    private List<Pair> buckets;

    public ArrayHashMap() {
        // Tömb inicializálása 100 vödörrel
        buckets = new ArrayList<>();
        for (int i = 0; i < 100; i++) {
            buckets.add(null);
        }
    }

    /* Hash függvény */
    private int hashFunc(int key) {
        int index = key % 100;
        return index;
    }

    /* Lekérdezési művelet */
    public String get(int key) {
        int index = hashFunc(key);
        Pair pair = buckets.get(index);
        if (pair == null)
            return null;
        return pair.val;
    }

    /* Hozzáadási művelet */
    public void put(int key, String val) {
        Pair pair = new Pair(key, val);
        int index = hashFunc(key);
        buckets.set(index, pair);
    }

    /* Törlési művelet */
    public void remove(int key) {
        int index = hashFunc(key);
        // null-ra állítja a törlés jelzéséhez
        buckets.set(index, null);
    }

    /* Összes kulcs-érték pár lekérése */
    public List<Pair> pairSet() {
        List<Pair> pairSet = new ArrayList<>();
        for (Pair pair : buckets) {
            if (pair != null)
                pairSet.add(pair);
        }
        return pairSet;
    }

    /* Összes kulcs lekérése */
    public List<Integer> keySet() {
        List<Integer> keySet = new ArrayList<>();
        for (Pair pair : buckets) {
            if (pair != null)
                keySet.add(pair.key);
        }
        return keySet;
    }

    /* Összes érték lekérése */
    public List<String> valueSet() {
        List<String> valueSet = new ArrayList<>();
        for (Pair pair : buckets) {
            if (pair != null)
                valueSet.add(pair.val);
        }
        return valueSet;
    }

    /* Hash tábla nyomtatása */
    public void print() {
        for (Pair kv : pairSet()) {
            System.out.println(kv.key + " -> " + kv.val);
        }
    }
}

public class array_hash_map {
    public static void main(String[] args) {
        /* Hash tábla inicializálása */
        ArrayHashMap map = new ArrayHashMap();

        /* Hozzáadási művelet */
        // Kulcs-érték pár (kulcs, érték) hozzáadása a hash táblához
        map.put(12836, "Xiao Ha");
        map.put(15937, "Xiao Luo");
        map.put(16750, "Xiao Suan");
        map.put(13276, "Xiao Fa");
        map.put(10583, "Xiao Ya");
        System.out.println("\nAfter adding is complete, hash table is\nKey -> Value");
        map.print();

        /* Lekérdezési művelet */
        // Kulcs beírása a hash táblába az érték lekéréséhez
        String name = map.get(15937);
        System.out.println("\nInput student ID 15937, query name " + name);

        /* Törlési művelet */
        // Kulcs-érték pár (kulcs, érték) eltávolítása a hash táblából
        map.remove(10583);
        System.out.println("\nAfter removing 10583, hash table is\nKey -> Value");
        map.print();

        /* Hash tábla bejárása */
        System.out.println("\nTraverse key-value pairs Key->Value");
        for (Pair kv : map.pairSet()) {
            System.out.println(kv.key + " -> " + kv.val);
        }
        System.out.println("\nTraverse keys only Key");
        for (int key : map.keySet()) {
            System.out.println(key);
        }
        System.out.println("\nTraverse values only Value");
        for (String val : map.valueSet()) {
            System.out.println(val);
        }
    }
}

File: hu/codes/python/chapter_hashing/array_hash_map.py
"""
File: array_hash_map.py
Created Time: 2022-12-14
Author: msk397 (machangxinq@gmail.com)
"""


class Pair:
    """Kulcs-érték pár"""

    def __init__(self, key: int, val: str):
        self.key = key
        self.val = val


class ArrayHashMap:
    """Tömbön alapuló hash tábla implementáció"""

    def __init__(self):
        """Konstruktor"""
        # Tömb inicializálása 100 vödörrel
        self.buckets: list[Pair | None] = [None] * 100

    def hash_func(self, key: int) -> int:
        """Hash függvény"""
        index = key % 100
        return index

    def get(self, key: int) -> str | None:
        """Lekérdezési művelet"""
        index: int = self.hash_func(key)
        pair: Pair = self.buckets[index]
        if pair is None:
            return None
        return pair.val

    def put(self, key: int, val: str):
        """Hozzáadás és frissítés"""
        pair = Pair(key, val)
        index: int = self.hash_func(key)
        self.buckets[index] = pair

    def remove(self, key: int):
        """Törlési művelet"""
        index: int = self.hash_func(key)
        # None-ra állítja a törlés jelölésére
        self.buckets[index] = None

    def entry_set(self) -> list[Pair]:
        """Az összes kulcs-érték pár lekérése"""
        result: list[Pair] = []
        for pair in self.buckets:
            if pair is not None:
                result.append(pair)
        return result

    def key_set(self) -> list[int]:
        """Az összes kulcs lekérése"""
        result = []
        for pair in self.buckets:
            if pair is not None:
                result.append(pair.key)
        return result

    def value_set(self) -> list[str]:
        """Az összes érték lekérése"""
        result = []
        for pair in self.buckets:
            if pair is not None:
                result.append(pair.val)
        return result

    def print(self):
        """Hash tábla kiírása"""
        for pair in self.buckets:
            if pair is not None:
                print(pair.key, "->", pair.val)


"""Fő kód"""
if __name__ == "__main__":
    # Hash tábla inicializálása
    hmap = ArrayHashMap()

    # Hozzáadás
    # Kulcs-érték pár (kulcs, érték) hozzáadása a hash táblához
    hmap.put(12836, "Xiao Ha")
    hmap.put(15937, "Xiao Luo")
    hmap.put(16750, "Xiao Suan")
    hmap.put(13276, "Xiao Fa")
    hmap.put(10583, "Xiao Ya")
    print("\nHozzáadás után a hash tábla:\nKulcs -> Érték")
    hmap.print()

    # Lekérdezés
    # Kulcs beírása a hash táblába az érték lekérdezéséhez
    name = hmap.get(15937)
    print("\nA 15937-es diákigazolvány alapján a név: " + name)

    # Törlés
    # Kulcs-érték pár (kulcs, érték) törlése a hash táblából
    hmap.remove(10583)
    print("\nA 10583-as törlése után a hash tábla:\nKulcs -> Érték")
    hmap.print()

    # Hash tábla bejárása
    print("\nKulcs-érték párok bejárása Kulcs->Érték")
    for pair in hmap.entry_set():
        print(pair.key, "->", pair.val)

    print("\nCsak kulcsok bejárása")
    for key in hmap.key_set():
        print(key)

    print("\nCsak értékek bejárása")
    for val in hmap.value_set():
        print(val)
